mathematical-companion/main.tex

```diff
@@ -0,0 +1,1300 @@
+errors can, in principle, be corrected.
+
+A \textbf{BCH code} is a specific type of error-correcting code. In a BCH
+code, codewords are constructed by encoding data as the coefficients of a
+large polynomial, then affixing some number of \textbf{checksum characters}.
+The checksum characters are chosen so that the full result is in
+a specific residue class modulo a \textbf{generator polynomial} $G(x)$.
+
+Given a potential codeword, there is a unique minimum-degree polynomial
+equivalent to this codeword modulo $G(x)$, which can be found by modular
+reduction. We refer to this reduced polynomial as the \textbf{residue}
+of the codeword.
+
+There are several properties of a BCH code that we will need:
+\begin{itemize}
+\item The \textbf{degree} of the code is the maximum degree of the roots of
+its generator polynomial. (Since the generator polynomial is not, in general,
+irreducible, this is \emph{not} the degree of the generator polynomial.)
+
+The degree of our BCH codes, and that of bech32, is 2. Degree-1 BCH codes are
+called \textbf{Reed-Solomon codes}.
+
+The degree indicates the dimension of the extension field you need to work
+in to do error correction. So for our codes, you would need to work in a
+quadratic extension of $\fttwo$, i.e.~$\ftttwo$.
+
+\item The \textbf{length} of the code is how long a coded message can be
+(including the checksum!) for the code to retain its error-correcting
+properties.
+
+The length $\ell$ can be computed as the smallest polynomial of the form
+$x^\ell - x$ that the generator divides. This can be seen by observing
+that if you have a message of this length, that the $\ell$th character
+will be interpreted as the coefficient of $x^{\ell-1} \equiv 1$ mod $G(x)$,
+meaning that it will just mask your first character rather than being
+checksummed independently.
+
+We have two codes -- the ``normal'' codex32 has a length of 93. For 512-bit
+seeds, we need more than 93 characters, so we use an alternate code of
+length 1023. The length of bech32 is also 1023 --- although through
+an exhaustive search, it was found to have better error detection properties
+than the algebra would suggest, up to length 71.
+
+\item The $m$-value, or \textbf{target residue}, is the specific residue that
+all codewords must must have modulo $G$. All $m$ values are equivalent, in
+the sense that it is easy to convert from one $m$ value to another (simply
+add the difference to the checksum characters of every codeword).
+
+But $m=0$ has the particularly bad property that any codeword can be extended
+by addition of an arbitrary number of 0s to get another valid codeword. (This
+highlights the fact that BCH codes are designed to handle only \emph{substitution}
+or \emph{erasure} errors, not insertions or deletions.) Other small values of
+$m$ have similar issues; bech32 was originally defined to have $m=1$ but later
+needed to be modified to bech32m for this reason. bech32m uses a large random
```

Comment:
not actually random.

Reply:
Do you know offhand how it was chosen? I may just say "random-looking" instead.

Reply:
Ah, sipa has a gist at https://gist.github.com/sipa/14c248c288c3880a3b191f978a34508e